rfd/0163-vnet.md

```diff
@@ -0,0 +1,971 @@
+    queryCluster -- No Match --> forward[Forward to default nameserver]
+    checkCustomDNS -- No --> forward[Forward to default nameserver]
+```
+
+Here is a seqence diagram for an example VNet session:
+
+```mermaid
+sequenceDiagram
+    participant Client
+    participant DNS_Resolver as DNS Resolver
+    participant VNet_DNS as VNet
+    participant Teleport as teleport.example.com
+    participant App_Service as App Service
+    participant API as API
+
+    Client->>DNS_Resolver: Ask for A record for api.teleport.example.com
+    DNS_Resolver->>DNS_Resolver: Check own cache
+    DNS_Resolver->>VNet_DNS: Ask for A record for api.teleport.example.com
+    VNet_DNS->>VNet_DNS: Check own cache
+    VNet_DNS->>Teleport: Ask if app "api" exists
+    Teleport->>VNet_DNS: Reply "api" exists
+    VNet_DNS->>VNet_DNS: Assign IP 100.64.0.2 to api.teleport.example.com
+    VNet_DNS->>DNS_Resolver: Return A record with IP 100.64.0.2
+    DNS_Resolver->>Client: Return A record with IP 100.64.0.2
+    Client->>VNet_DNS: Open TCP connection with 100.64.0.2
+    VNet_DNS->>Teleport: Proxy connection to cluster
+    Teleport->>App_Service: Proxy connection to app service
+    App_Service->>API: Proxy connection to 127.0.0.1:8888
+```
+
+### IP range
+
+#### IPv4
+
+VNet IP addresses are only ever used locally on a single device for applications
+connecting to the virtual TUN interface.
+Because these IPs do not need to be routable or advertised outside of the client
+machine, this gives us with essentially two options:
+1. Use a subset of one of the private IP ranges `10/8`, `172.16/12`, or `192.168/16`
+   as defined in [RFC 1918](https://datatracker.ietf.org/doc/html/rfc1918).
+2. Use the carrier-grade NAT shared address space `100.64/10` defined in
+   [RFC 6598](https://datatracker.ietf.org/doc/html/rfc6598).
+
+There is some prior art for using the carrier-grade NAT (CGNAT) shared address
+space for our intended purpose, it is already in use by other VPN-replacement
+apps such as [tailscale](https://tailscale.com/) and [Enclave](https://enclave.io/).
+
+VNet will use the CGNAT range by default, but we will make it configurable at
+the cluster level in case this conflicts with any customer's network.
+
+The primary reason for using the CGNAT range is to avoid IP and route conflicts
+with the RFC 1918 private address spaces that are very commonly used in
+enterprises already.
+There are 4 million IP addresses in this range so exhaustion won't be a concern.
+This range can be overriden with a cluster-wide configuration.
+
+One downside of using the CGNAT range is that the Chromium web browser does not
+currently consider it a private address space for the purposes of CORS
+protections against DNS rebinding attacks.
+See [DNS rebinding](#dns-rebinding).
+
+Configuring a custom IP range at the cluster level would look like:
+
+```yaml
+version: v1
+kind: vnet
+metadata:
+  name: vnet
+spec:
+  cidr_range: "100.64.0.0/10"
```

Comment:
Not sure, could we leave that in the hands of the teleport admin configuring this? Or we could make an allowlist

Reply:
My concern primarily comes from the potential non-deterministic handling in the routing table on a possible conflict.

If we override the public route to force the request internal to a Teleport resource that was unexpected, that seems potentially concerning.  But it's potentially more concerning if the public route were to be followed (This is particularly concerning without https / TLS cert validation being discussed below, since it could allow a takeover of the unvalidated but expected internal resource).

I think having a whitelist of CIDR's which are allowed would be safest.